graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java

```diff
@@ -0,0 +1,146 @@
+    this.entity = entity;
+    this.field = field;
+    this.entities = entities;
+    this.context = context;
+  }
+
+  @Override
+  IncrementModel build() throws SkipException {
+
+    Optional<Directive> directive = DirectiveHelper.getDirective(CqlDirectives.INCREMENT, argument);
+
+    boolean prepend =
+        directive
+            .flatMap(d -> DirectiveHelper.getBooleanArgument(d, "prepend", context))
+            .orElse(PREPEND_DEFAULT);
+
+    validate(field, prepend);
+    return new IncrementModel(field, prepend, argument.getName());
+  }
+
+  protected void validate(FieldModel field, boolean prepend) throws SkipException {
+    if (field.isPartitionKey() || field.isClusteringColumn()) {
+      invalidMapping(
+          "Operation %s: directive %s is not supported for partition/clustering key field %s.",
+          operationName, CqlDirectives.INCREMENT, field.getGraphqlName());
+      throw SkipException.INSTANCE;
+    } else {
+      checkValidForRegularColumn(field, prepend);
+    }
+  }
+
+  private void checkValidForRegularColumn(FieldModel field, boolean prepend) throws SkipException {
+
+    Type<?> fieldInputType =
+        toInput(field.getGraphqlType(), argument, entity, field, entities, operationName);
+    // if it is non-list type
+    if (fieldInputType instanceof TypeName) {
+      validateCounter((TypeName) fieldInputType);
+      validatePrepend(field, prepend);
+    } else {
+      // otherwise it must be a list
+      checkArgumentIsListOf(argument, entity, entities, operationName, field);
+    }
+  }
+
+  private void validateCounter(TypeName fieldInputType) throws SkipException {
+    String typeName = fieldInputType.getName();
+    if (typeName.equalsIgnoreCase(COUNTER_TYPE_NAME)) {
+      // counter graph-ql field can be BIGINT or INT
+      checkArgumentIsAnyOfTypes(
+          Arrays.asList(CqlScalar.BIGINT.getGraphqlType(), Scalars.GraphQLInt));
+    }
+  }
+
+  private void validatePrepend(FieldModel field, boolean prepend) throws SkipException {
+    if (prepend) {
+      Type<?> fieldInputType =
+          toInput(field.getGraphqlType(), argument, entity, field, entities, operationName);
+      if (!(fieldInputType instanceof ListType)) {
+        invalidMapping(
+            "Operation %s: the %s directive with prepend = true cannot be used with argument %s "
+                + "because it is not a list",
+            operationName, CqlDirectives.INCREMENT, argument.getName());
+        throw SkipException.INSTANCE;
+      }
+    }
+  }
+
+  protected void checkArgumentIsAnyOfTypes(List<GraphQLScalarType> scalarTypes)
+      throws SkipException {
+
+    Type<?> argumentType = TypeHelper.unwrapNonNull(argument.getType());
+
+    boolean hasProperType =
+        scalarTypes.stream()
+            .anyMatch(
+                type -> {
+                  if (argumentType instanceof TypeName) {
+                    return type.getName().equals(((TypeName) argumentType).getName());
+                  }
```

Comment:
Nit: you can extract `argumentType.getName()` outside of the loop.

Reply:
To use it in lambda we would still need to copy it:
```
 String argumentTypeName = null;
    if (argumentType instanceof TypeName) {
      argumentTypeName = ((TypeName) argumentType).getName();
    }
    String finalArgumentTypeName = argumentTypeName;
```
so I am not sure if extracting it is a good choice.

Reply:
Addressed in #e9b4f103